internal/locks/locks.go

```diff
@@ -0,0 +1,89 @@
+package locks
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/speakeasy-api/speakeasy/internal/singleton"
+)
+
+// Package locks provides utilities for inter-process synchronization.
+
+// CLIUpdateMutex provides file-based mutual exclusion between processes.
+// It uses file locking to ensure that only one process can perform CLI updates at a time.
+// The lock is automatically released if the holding process dies.
+//
+// This implementation is cross-platform compatible:
+//   - Linux/macOS: Uses flock syscall (https://linux.die.net/man/2/flock)
+//   - Windows: Uses LockFileEx API (https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex)
+//
+// The mutex can be used to coordinate CLI update operations between multiple
+// concurrent processes to prevent race conditions during installation or updates.
+type CLIUpdateMutex struct {
+	Opts
+	mu *flock.Flock
+}
+
+type Opts struct {
+	Name string
+}
+
+func DefaultOpts() Opts {
+	return Opts{Name: "speakeasy.lock"}
+}
+
+func new(o Opts) *CLIUpdateMutex {
+	mu := flock.New(filepath.Join(os.TempDir(), o.Name))
+
+	return &CLIUpdateMutex{Opts: o, mu: mu}
+}
+
+// CLIUpdateLockWithOpts creates a new inter-process mutex with a custom config.
+var CLIUpdateLockWithOpts = singleton.NewWithOpts(func(o Opts) *CLIUpdateMutex {
+	return new(o)
```

Comment:
Nit: Name should not have a default - this is the `speakeasy-cli-update-mutex.lock`